In [2]:
import pandas as pd
import numpy as np

In [3]:
#dataset
url = 'https://raw.githubusercontent.com/Benura2020/LaptopPricePredictor/main/Data/laptop_price.csv'
data = pd.read_csv(url, encoding='latin-1')

## Data Analysis

In [4]:
# amount of data 
data.shape

(1303, 12)

In [5]:
# check null values
data.isnull().sum()

laptop_ID           0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   object 
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight            1303 non-null   object 
 11  Price_euros       1303 non-null   float64
dtypes: float64(2), int64(1), object(9)
memory usage: 122.3+ KB


In [7]:
data.head(3)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   

                     ScreenResolution                         Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz  8GB   
1                            1440x900        Intel Core i5 1.8GHz  8GB   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz  8GB   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS  1.37kg      1339.69  
1        Intel HD Graphics 6000  macOS  1.34kg       898.94  
2         Intel HD Graphics 620  No OS  1.86kg       575.00

'laptop_ID', 'Inches' are not required to predict the price /
'Inches', 'Price_euros' - no need to convert /
'Ram' - remove GB /
'Weight' - remove kg

In [8]:
# modify Ram column
data['Ram'] = data['Ram'].str.replace('GB','').astype('int32')

In [9]:
# modify Weight column
data['Weight'] = data['Weight'].str.replace('kg','').astype('float32')

In [10]:
data.head(3)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   

                     ScreenResolution                         Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz    8   
1                            1440x900        Intel Core i5 1.8GHz    8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz    8   

                            Gpu  OpSys  Weight  Price_euros  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69  
1        Intel HD Graphics 6000  macOS    1.34       898.94  
2         Intel HD Graphics 620  No OS    1.86       575.00

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1303 entries, 0 to 1302
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         1303 non-null   int64  
 1   Company           1303 non-null   object 
 2   Product           1303 non-null   object 
 3   TypeName          1303 non-null   object 
 4   Inches            1303 non-null   float64
 5   ScreenResolution  1303 non-null   object 
 6   Cpu               1303 non-null   object 
 7   Ram               1303 non-null   int32  
 8   Gpu               1303 non-null   object 
 9   OpSys             1303 non-null   object 
 10  Weight            1303 non-null   float32
 11  Price_euros       1303 non-null   float64
dtypes: float32(1), float64(2), int32(1), int64(1), object(7)
memory usage: 112.1+ KB


Company column

In [12]:
# Calculate value counts for each company
company_counts = data['Company'].value_counts()

In [13]:
company_counts

Company
Dell         297
Lenovo       297
HP           274
Asus         158
Acer         103
MSI           54
Toshiba       48
Apple         21
Samsung        9
Razer          7
Mediacom       7
Microsoft      6
Xiaomi         4
Vero           4
Chuwi          3
Google         3
Fujitsu        3
LG             3
Huawei         2
Name: count, dtype: int64

Create a new column for the value count is less than 10

In [14]:
# Get list of companies with counts less than or equal to 10
companies_less_than_10 = company_counts[company_counts <= 10].index.tolist()

print(companies_less_than_10)

['Samsung', 'Razer', 'Mediacom', 'Microsoft', 'Xiaomi', 'Vero', 'Chuwi', 'Google', 'Fujitsu', 'LG', 'Huawei']


In [15]:
# for count less than 10, create a new column

def add_company(input):
    if input in companies_less_than_10:
        return 'Other' 
    else:
        return input
data['Company'] = data['Company'].apply(add_company)

In [16]:
data['Company'].value_counts()

Company
Dell       297
Lenovo     297
HP         274
Asus       158
Acer       103
MSI         54
Other       51
Toshiba     48
Apple       21
Name: count, dtype: int64

Product column

In [17]:
data['Product'].value_counts()

Product
XPS 13                                     30
Inspiron 3567                              29
250 G6                                     21
Legion Y520-15IKBN                         19
Vostro 3568                                19
                                           ..
15-bw007nv (A10-9620P/6GB/128GB/Radeon      1
SmartBook 130                               1
A541NA-GO342 (N3350/4GB/500GB/Linux)        1
17-X047na (i3-6006U/8GB/1TB/W10)            1
V330-15IKB (i5-8250U/4GB/500GB/FHD/W10)     1
Name: count, Length: 618, dtype: int64

In [18]:
len(data['Product'].value_counts())

618

There are 618 categories in Product column. So we can consider there will be a little effect to Price from Product column. want to remove Product column

TypeName column

In [19]:
data['TypeName'].value_counts()

TypeName
Notebook              727
Gaming                205
Ultrabook             196
2 in 1 Convertible    121
Workstation            29
Netbook                25
Name: count, dtype: int64

ScreenResolution

In [20]:
data['ScreenResolution'].value_counts()

ScreenResolution
Full HD 1920x1080                                507
1366x768                                         281
IPS Panel Full HD 1920x1080                      230
IPS Panel Full HD / Touchscreen 1920x1080         53
Full HD / Touchscreen 1920x1080                   47
1600x900                                          23
Touchscreen 1366x768                              16
Quad HD+ / Touchscreen 3200x1800                  15
IPS Panel 4K Ultra HD 3840x2160                   12
IPS Panel 4K Ultra HD / Touchscreen 3840x2160     11
4K Ultra HD / Touchscreen 3840x2160               10
4K Ultra HD 3840x2160                              7
Touchscreen 2560x1440                              7
IPS Panel 1366x768                                 7
IPS Panel Quad HD+ / Touchscreen 3200x1800         6
IPS Panel Retina Display 2560x1600                 6
IPS Panel Retina Display 2304x1440                 6
Touchscreen 2256x1504                              6
IPS Panel Touchscreen 2560x14

In [21]:
len(data['ScreenResolution'].value_counts())

40

It has some technologies. Modify ScreenResolution column and create new 2 columns for Touch screen and Ips. If it has Touch screen or Ips, it store 1.Else it will store 0

In [22]:
data['Touchscreen'] = data['ScreenResolution'].apply(lambda x:1 if 'Touchscreen' in x else 0)
data['Ips'] = data['ScreenResolution'].apply(lambda x:1 if 'IPS' in x else 0)

So now we can remove ScreenResolution column

In [23]:
data.head(3)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   

                     ScreenResolution                         Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz    8   
1                            1440x900        Intel Core i5 1.8GHz    8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1  
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0  
2         Intel HD Graphics 620  No OS    1.86       575.00            0    0

Cpu column

In [24]:
data['Cpu'].value_counts()

Cpu
Intel Core i5 7200U 2.5GHz       190
Intel Core i7 7700HQ 2.8GHz      146
Intel Core i7 7500U 2.7GHz       134
Intel Core i7 8550U 1.8GHz        73
Intel Core i5 8250U 1.6GHz        72
                                ... 
Intel Core M M3-6Y30 0.9GHz        1
AMD A9-Series 9420 2.9GHz          1
Intel Core i3 6006U 2.2GHz         1
AMD A6-Series 7310 2GHz            1
Intel Xeon E3-1535M v6 3.1GHz      1
Name: count, Length: 118, dtype: int64

In [25]:
len(data['Cpu'].value_counts())

118

Can't use one hot encoding for all categories. Can divide into categories without considering speed

In [26]:
#check first 3 words to identify cpu without speed
data['cpu_name'] = data['Cpu'].apply(lambda x:" ".join(x.split()[0:3]))

In [27]:
data['cpu_name'].value_counts()

cpu_name
Intel Core i7               527
Intel Core i5               423
Intel Core i3               136
Intel Celeron Dual           80
Intel Pentium Quad           27
Intel Core M                 19
AMD A9-Series 9420           12
Intel Celeron Quad            8
AMD A6-Series 9220            8
AMD A12-Series 9720P          7
Intel Atom x5-Z8350           5
AMD A8-Series 7410            4
Intel Atom x5-Z8550           4
Intel Pentium Dual            3
AMD A9-Series 9410            3
AMD Ryzen 1700                3
AMD A9-Series A9-9420         2
AMD A10-Series 9620P          2
Intel Atom X5-Z8350           2
AMD E-Series E2-9000e         2
Intel Xeon E3-1535M           2
Intel Xeon E3-1505M           2
AMD E-Series 7110             2
AMD A10-Series 9600P          2
AMD A6-Series A6-9220         2
AMD A10-Series A10-9620P      2
AMD Ryzen 1600                1
Intel Atom x5-Z8300           1
AMD E-Series E2-6110          1
AMD FX 9830P                  1
AMD E-Series E2-9000          1

Divide categories as this. Intel core i7, Intel core i5, Intel core i3, AMD and other

In [28]:
def set_processor(name):
    if name == 'Intel Core i7' or name == 'Intel Core i5' or name == 'Intel Core i3':
        return name
    else:
        if name.split()[0] == 'AMD':
            return 'AMD'
        else:
            return 'Other'

data['cpu_name'] = data['cpu_name'].apply(set_processor)

In [29]:
data['cpu_name'].value_counts()

cpu_name
Intel Core i7    527
Intel Core i5    423
Other            155
Intel Core i3    136
AMD               62
Name: count, dtype: int64

So now we can drop cpu column

Gpu column

In [30]:
data['Gpu'].value_counts()

Gpu
Intel HD Graphics 620      281
Intel HD Graphics 520      185
Intel UHD Graphics 620      68
Nvidia GeForce GTX 1050     66
Nvidia GeForce GTX 1060     48
                          ... 
AMD Radeon R5 520            1
AMD Radeon R7                1
Intel HD Graphics 540        1
AMD Radeon 540               1
ARM Mali T860 MP4            1
Name: count, Length: 110, dtype: int64

In [31]:
len(data['Gpu'].value_counts())

110

In [32]:
#check first word to identify gpu type 
data['gpu_name'] = data['Gpu'].apply(lambda x:" ".join(x.split()[0:1]))

In [33]:
data['gpu_name'].value_counts()

gpu_name
Intel     722
Nvidia    400
AMD       180
ARM         1
Name: count, dtype: int64

Can remove ARM gpu row

In [34]:
data.shape

(1303, 16)

In [35]:
data = data[data['gpu_name'] != 'ARM']  # filtering out the rows where the 'gpu_name' is not equal to 'ARM'

In [36]:
data.shape

(1302, 16)

In [37]:
data.head(3)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   

                     ScreenResolution                         Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz    8   
1                            1440x900        Intel Core i5 1.8GHz    8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640  macOS    1.37      1339.69            0    1   
1        Intel HD Graphics 6000  macOS    1.34       898.94            0    0   
2         Intel HD Graphics 620  No OS    1.86       575.00            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel  
2  Intel Core i5    Intel

OpSys column

In [38]:
data['OpSys'].value_counts()

OpSys
Windows 10      1072
No OS             66
Linux             62
Windows 7         45
Chrome OS         26
macOS             13
Mac OS X           8
Windows 10 S       8
Android            2
Name: count, dtype: int64

Divide categories as this. Winows, macOS or Mac OS, Linux and other

In [39]:
def set_os(name):
    if name == 'Windows 10' or name == 'Windows 7' or name == 'Windows 10 S':
        return 'Windows'
    elif name == 'macOS' or name == 'Mac OS X':
        return 'Mac'
    elif name == 'Linux':
        return name
    else:
        return 'Other' 
        
data['OpSys'] = data['OpSys'].apply(set_os)

In [40]:
data['OpSys'].value_counts()

OpSys
Windows    1125
Other        94
Linux        62
Mac          21
Name: count, dtype: int64

In [41]:
data.head(3)

laptop_ID Company      Product   TypeName  Inches  \
0          1   Apple  MacBook Pro  Ultrabook    13.3   
1          2   Apple  Macbook Air  Ultrabook    13.3   
2          3      HP       250 G6   Notebook    15.6   

                     ScreenResolution                         Cpu  Ram  \
0  IPS Panel Retina Display 2560x1600        Intel Core i5 2.3GHz    8   
1                            1440x900        Intel Core i5 1.8GHz    8   
2                   Full HD 1920x1080  Intel Core i5 7200U 2.5GHz    8   

                            Gpu  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0  Intel Iris Plus Graphics 640    Mac    1.37      1339.69            0    1   
1        Intel HD Graphics 6000    Mac    1.34       898.94            0    0   
2         Intel HD Graphics 620  Other    1.86       575.00            0    0   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel  
2  Intel Core i5    Intel

Delete unwanted columns (laptop_ID, Inches, Product, ScreenResolution, Cpu, Gpu)

In [42]:
data = data.drop(columns=['laptop_ID', 'Inches', 'Product', 'ScreenResolution', 'Cpu', 'Gpu'])

In [43]:
data.head(5)

Company   TypeName  Ram  OpSys  Weight  Price_euros  Touchscreen  Ips  \
0   Apple  Ultrabook    8    Mac    1.37      1339.69            0    1   
1   Apple  Ultrabook    8    Mac    1.34       898.94            0    0   
2      HP   Notebook    8  Other    1.86       575.00            0    0   
3   Apple  Ultrabook   16    Mac    1.83      2537.45            0    1   
4   Apple  Ultrabook    8    Mac    1.37      1803.60            0    1   

        cpu_name gpu_name  
0  Intel Core i5    Intel  
1  Intel Core i5    Intel  
2  Intel Core i5    Intel  
3  Intel Core i7      AMD  
4  Intel Core i5    Intel

One hot encoding

In [44]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1302 entries, 0 to 1302
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Company      1302 non-null   object 
 1   TypeName     1302 non-null   object 
 2   Ram          1302 non-null   int32  
 3   OpSys        1302 non-null   object 
 4   Weight       1302 non-null   float32
 5   Price_euros  1302 non-null   float64
 6   Touchscreen  1302 non-null   int64  
 7   Ips          1302 non-null   int64  
 8   cpu_name     1302 non-null   object 
 9   gpu_name     1302 non-null   object 
dtypes: float32(1), float64(1), int32(1), int64(2), object(5)
memory usage: 101.7+ KB


One hot encode - Company, TypeName, OpSys, cpu_name, gpu_name'

In [45]:
# there will no changes in Ram, Weight, Price_euros, Touchscreen, ips columns
data = pd.get_dummies(data, dtype=int)
# columns = ['Company', 'TypeName', 'OpSys', 'cpu_name', 'gpu_name']

In [46]:
data.head()

Ram  Weight  Price_euros  Touchscreen  Ips  Company_Acer  Company_Apple  \
0    8    1.37      1339.69            0    1             0              1   
1    8    1.34       898.94            0    0             0              1   
2    8    1.86       575.00            0    0             0              0   
3   16    1.83      2537.45            0    1             0              1   
4    8    1.37      1803.60            0    1             0              1   

   Company_Asus  Company_Dell  Company_HP  ...  OpSys_Other  OpSys_Windows  \
0             0             0           0  ...            0              0   
1             0             0           0  ...            0              0   
2             0             0           1  ...            1              0   
3             0             0           0  ...            0              0   
4             0             0           0  ...            0              0   

   cpu_name_AMD  cpu_name_Intel Core i3  cpu_name_Intel Core i5  \
0             0                       0                       1   
1             0                       0                       1   
2             0                       0                       1   
3             0                       0                       0   
4             0                       0                       1   

   cpu_name_Intel Core i7  cpu_name_Other  gpu_name_AMD  gpu_name_Intel  \
0                       0               0             0               1   
1                       0               0             0               1   
2                       0               0             0               1   
3                       1               0             1               0   
4                       0               0             0               1   

   gpu_name_Nvidia  
0                0  
1                0  
2                0  
3                0  
4                0  

[5 rows x 32 columns]

In [47]:
data.shape

(1302, 32)

## Model Building

In [48]:
#features
X = data.drop('Price_euros', axis=1)
y = data['Price_euros']

In [49]:
#split dataset for training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25)

In [50]:
X_train.shape, X_test.shape

((976, 31), (326, 31))

This is a regression problem

We check what is the most suitable algorithm by trainig these algorithms - Linera regression, lasso regression, decision tree, random forest

In [59]:
# function to check model accuracy of each models
def model_accuracy_check(model):
    model.fit(X_train, y_train)
    model_accuracy = model.score(X_test, y_test)
    print(str(model)+ '--->'+str(model_accuracy)) 

### Linear Regression

In [92]:
from sklearn.linear_model import LinearRegression
linearRegression = LinearRegression()

### Lasso Regression

In [93]:
from sklearn.linear_model import Lasso
lasso = Lasso()

### Decision Tree Regressor

In [94]:
from sklearn.tree import DecisionTreeRegressor
decisionTree = DecisionTreeRegressor()

### Random Forest Regressor

In [95]:
from sklearn.ensemble import RandomForestRegressor
randomforest = RandomForestRegressor()

In [110]:
# compare models
model_accuracy_check(linearRegression)
model_accuracy_check(lasso)
model_accuracy_check(decisionTree)
model_accuracy_check(randomforest)

LinearRegression()--->0.7256696110274767
Lasso()--->0.7250650470865317
DecisionTreeRegressor()--->0.5232695428684515
RandomForestRegressor()--->0.7502455109346646


So we can select RandomForestRegressor() to train the final model

### Hyperparameter tuning

In [107]:
from sklearn.model_selection import GridSearchCV

# use 2 parameters
parameters = {'n_estimators': [10, 50, 100],
              'criterion': ['squared_error', 'absolute_error', 'poisson']}

#creating a grid object
grid_obj = GridSearchCV(estimator=randomforest, param_grid=parameters)

# train 
grid_fit = grid_obj.fit(X_train, y_train)

# create a model using most suitable parameters
best_model = grid_fit.best_estimator_
best_model

RandomForestRegressor(criterion='absolute_error', n_estimators=50)

In [108]:
best_model.score(X_test, y_test)

0.7649377925459466

### Save model

In [114]:
import pickle
with open('price_predictor.pickle', 'wb') as file:
    pickle.dump(best_model, file)